### PR TITLE
Added Create Annotated Table section using the algorithm outlined in #193

### LINKED ISSUE #193: Documenting the processing steps for metadata
At the telco on [2015-02-04](http://www.w3.org/2015/02/04-csvw-minutes.html) the idea of documenting the processing steps came up. The [section on parsing](http://www.w3.org/TR/tabular-data-model/#parsing) is not the one, indeed, that section concentrates on the _parsing_ of a CSV file, provided the various dialect parameters are given, and it does not touch upon the exact processing steps to be followed to generate the right metadata, for example. I think such a section should be part of one of our documents, possibly as a normative section. 

I have made a first attempt below to describe the process; obviously to be refined. But that is my understanding of the current documents, and I may have of course misunderstood some things. But if I did, while doing the implementations, than maybe others will, too, which is all the more of a reason to document it...

TL;DR: the result is, unfortunately, complex. I would say: too complex, actually. 
# Processing steps
## If the process starts with the URL of a CSV file:
1. Get the raw CSV content from the file through an `HTTP GET` (or access it from the local file system). Store the Link Header from the return, too.
2. Get hold of the following metadata objects (see the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document): 
   - metadata supplied by the user of the implementation that is processing the tabular data (`UM`) 
   - metadata in a document linked to using a Link Header associated with the tabular data file (`LM`)
   - file-specific metadata in a document located based on the location of the tabular data file (`FM`)
   - directory-specific metadata in a document located based on the location of the tabular data file (`DM`)
3. Create a merge of these metadata `M1 = merge(UM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification
4. If `M1` does not include a [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions), add a default (_this still has to be defined_)
5. Possibly modify the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` if the `;header=absent` is set in the `Content-type` of the returned Link Header (see the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#headers)) 
6. Parse the CSV data using the [Dialect Description](http://www.w3.org/TR/tabular-metadata/#dialect-descriptions) of `M1` . An informal description on the effects of the various dialect parameter and the parsing itself is part of the [Tabular Model Document](http://www.w3.org/TR/tabular-data-model/#parsing). 
7. Retrieve the metadata embedded within the tabular data file (`EM`) (see again the [section on locating metadata](http://www.w3.org/TR/tabular-data-model/#locating-metadata) in the tabular data model document). Note that this specification does not define the syntax of that embedded metadata, if any, except if `EM` consists of a single header row (providing `title` values for the corresponding `column` objects)
8. Determine the default Metadata (`DM`) (_this still has to be formally defined and described_)
9. Create a merge of `M = merge(DM, UM, EM, LM, FM, DM)` using the [merge algorithm](http://www.w3.org/TR/tabular-metadata/#merging-metadata) defined in the Tabular Metadata specification
10. The final Annotated Tabular Data consists of the parsed raw data, possibly disregarding the rows according to the Dialect Description, with `M` as the final metadata.
11. If required by the user, use the raw data and `M` to generate either JSON or RDF
## If the process starts with the URL of a Metadata file:

T.B.D.
# Comments:
- step no. 9 is awkward. Of course, it can be optimized by storing `merge(LM,FM,DM)` separately, but the fact that the priority order places `UM` ahead of `EM` means that such a re-run becomes necessary
- same for DM (steps nos. 8 and 9): it is unclear where exactly in the process `DM` comes in and is also part of the re-merge.
- much of the complication comes, in my view, of the fact that step no. 6, i.e., the parsing itself, is _dependent_ on the (dialect) data stored in the metadata. I wonder whether separating the dialect information completely from the metadata, either to be retrieved separately, or left for the implementation, might not be a better approach. This comes to my question I was asking before: is the Dialect information in the metadata _prescriptive_ or _descriptive_; the current model is the former (and leads to a complexity), whereas if it was the latter, that would simplify metadata processing. (I.e., the processor would have to extend `M` with the dialect data _once the parsing occurred_ but the final merge of all the metadata could then be delayed and done only once.)

(Note that this is also related to issue #145, see https://github.com/w3c/csvw/issues/145#issuecomment-68828919, and to issue #144, see https://github.com/w3c/csvw/issues/144#issuecomment-69032693)

### PR DESCRIPTION
**Pull request by gkellogg**
This fixes #193.

### CODE REVIEW

**Comment by iherman** [comment]
See my comment at https://github.com/w3c/csvw/issues/193#issuecomment-75913028

Ivan

> On 25 Feb 2015, at 06:43 , Gregg Kellogg notifications@github.com wrote:
> 
> This fixes #193.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/w3c/csvw/pull/259
> 
> Commit Summary
> 
>   • Added Create Annotated Table section using the algorithm outlined in #193.
> File Changes
> 
>   • M syntax/index.html (75)
> Patch Links:
> 
>   • https://github.com/w3c/csvw/pull/259.patch
>   • https://github.com/w3c/csvw/pull/259.diff
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by JeniT** [comment]
I edited the description a bit to reflect my understanding of the processing. @gkellogg and @iherman please check.

File: syntax/index.html [hunk 1]
```diff
@@ -349,6 +349,33 @@
...
+          <li>If the tabular data file was retrieved with <code>Content-Type</code> including the <code>header=absent</code> parameter set <code>header</code> to <code>false</code> in <code>DD</code>.</li>
+          <li>Parse the tabular data file using <code>DD</code> to create a basic tabular data model (<code>T</code>) with core annotations. Extract embedded metadata (<code>EM</code>), for example from the <a>header line</a>. <span class="note">Note: This specification provides a non-normative definition of how embedded metadata may be extracted from tabular data files in <a href="#parsing-tabular-data" class="sectionRef"></a>.</span></li>
```
Comment (iherman) [inline syntax/index.html]: The note on embedded metadata may be misunderstood as if there was a more explicit informative definition on extracting embedded metadata in the parsing section, though there is not any (which is the way it should be). What about something like

[[[
Note: This specification provides non-normative definition on parsing tabular data which may also include the extraction of embedded metadata in <a href="#parsing-tabular-data" class="sectionRef"></a>. However, no further definition for the format and syntax on embedded metadata is provided. 
]]]

or something like that.

File: syntax/index.html [hunk 2]
```diff
@@ -349,6 +349,33 @@
...
+          <li>Create the merged metadata <code>M = merge(UM, EM, LM, FM, DM)</code> using the <a href="http://w3c.github.io/csvw/metadata/#dfn-merged" class="externalDFN">merge algorithm</a> defined in [[!tabular-metadata]].</li>
```
Comment (iherman) [inline syntax/index.html]: Following the discussions on parsing, I wonder whether there is a need to update the Dialect Description in M based on parsing. For example, the parser may detect a different line ending and may return that information as part of the processing; in which case the final Dialect Description in M should probably be updated. (Although this is a rare occasion indeed.)

**Comment by JeniT** [comment]
@iherman take a look now; I've rephrased the note and around the use of the dialect description

**Comment by iherman** [comment]
Thanks @JeniT, I like it!

Ivan

> On 01 Mar 2015, at 09:53 , Jeni Tennison notifications@github.com wrote:
> 
> @iherman take a look now; I've rephrased the note and around the use of the dialect description
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704
